### DISCUSSION TITLE
FW v6 release - CLOSED

### DISCUSSION THREAD

**Discussion by UltrashRicco** [astlouys/Pico-Green-Clock#3] [General]
Hi Alain, thank you for releasing FW v6, I am currently testing it. I am attaching my custom version.
[Pico-Clock-Green-V6-Eric.zip](https://github.com/astlouys/Pico-Green-Clock/files/9928000/Pico-Clock-Green-V6-Eric.zip)

Looks like the conditional compile issues were fixed, great job! 

With the QUICK_START conditional compile option, it is now much easier to customize behaviour at boot time, thanks!

Thanks for implementing my suggestions: °C VS °c, and automatically light up the "AutoLight" LEDs if auto-brightness is activated.

I disabled Daylight Time Savings for now, but I may try and come up with a version that works in Europe (the dates are different from those in North America).

**Comment by astlouys** [comment]
Hi Eric,

(It's "André"...) Yes, many goodies / fixes that I added in Version 6.00 were based on comments received from users.

I'm working on a Version 7.00 during my spare time... I spent many hours in the last weeks trying to figure out how to make the DHT22 work in a decent way (I'm using it for indoor temperature and humidity since I use a BME280 for outdoor data)...  I finally decided to offload its tasks to the Pico's second core and it works fine now. So, if you can provide me with specific details about DST in Spain, I may take a look and add it to the same V7.00. The release may be too late for next change, but as we say: "better later than never". I tried to figure out a way to easily add an hour during the setup, but I didn't like any of the solutions I came up with. They make the setting counter-intuitive (worst than day-of-week with alarms!).

(P.S. I realize that your clock is scrolling data quite fast, and often !! - smile)

**Reply by UltrashRicco** [reply]
Sorry André! :) Yes I like to keep an eye on the temperature from time to time, and if you miss the scroll with a low scrolling frequency, too bad, you need to wait and keep an eye on the clock until the next cycle! Also I like it to scroll faster indeed.

This is the info that I could find about DST in Europe VS NA. BTW, we already changed hours last weekend! https://en.wikipedia.org/wiki/Daylight_saving_time
"Since 1996, European Summer Time has been observed from the last Sunday in March to the last Sunday in October; previously the rules were not uniform across the European Union. Starting in 2007, most of the United States and Canada observed DST from the second Sunday in March to the first Sunday in November, almost two-thirds of the year."

So, it would not be as easy as counting how many Sundays have elapsed in the month, you need to know how many sundays there are in March/October of the current year, and then check whether it is the last sunday of the month... or somehow calculate the date at which the change is to happen.
I will ckeck the logic in your code and see if I can figure something out.
Also, not sure I understand the values you declared for DST_LO_LIMIT, DST_NONE and DST_HI_LIMIT, but I think I understand it has to do with the display while setting the value from the actual clock buttons.

**Reply by UltrashRicco** [reply]
![image](https://user-images.githubusercontent.com/54411674/199738134-137658ef-0b69-4290-b5b2-4763a9ab4858.png)

I guess it should be possible to do this very same calculation for the whole month, and increment a "TotalSundayCounter", then reuse the original code, and when SundayCounter equals TotalSundayCounter, you know you are at (or past) the last Sunday of the month.

**Comment by astlouys** [comment]
I need to take a closer look, but at first glance, I think that I can greatly simplify the algorithm and come up with something much simpler than what I've already done (and much easier to adapt to different countries if ever the case...)

I'll keep you informed. Do you volunteer to make testing for Europe implementation ?

**Reply by UltrashRicco** [reply]
Sure! I would love to code it myself but I do not think I am able to do that. I might be able to suggest part of the logic though. :)
I would be willing to test for sure. It would be as easy as setting the time to a few minutes before the change is bound to happen, and check whether it actually happens.

Europe would not be too hard to implement, I think. Not sure how many other zones are to be taken into consideration, but as Olav mentionned, this could quickly become a can of worms! :) https://en.wikipedia.org/wiki/Daylight_saving_time_by_country

Europe voted for DST to be dropped back in 2018. 86% of the voting European citizens decided agains keeping DST.
2021 was to implement it, but it did not happen due to the global COVID situation, etc. The topic was not addressed again yet since then! https://en.wikipedia.org/wiki/Daylight_saving_time_by_country#European_Union

**Reply by astlouys** [reply]
Eric,

I spent a few hours in the last days working on a new algorithm for Daylight Saving Time / Summer Time...

I still have some more tests to do, but I believe it will cover most (if not all) countries listed in the wiki on the subject. I'll let you perform a few tests for European Union (and other countries if you want !) and hopefully users from other countries will join us and will give it a try for the parameters used in their respective countries.

For the sake of completeness, the tests should cover two different things:

1) When the clock is powered up, real-time clock has been updated internally with the battery backed-up RTC IC. However, this real-time clock has no clue about DST / Summer Time status. So, based on the DST setting (that the user previously configured and kept in Pico's flash), the clock will determine if we are currently during the DST / Summer Time active period or not.

2) When we reach the day of the year when DST / Summer Time status changes, we should make sure it is properly handled by the clock.

In fact, the first item above is much more work to do than the second one, but I believe I covered both...

Maybe we should open another discussion thread about Daylight Saving Time / Summer Time...

P. S. I'll need a few more days to publish the update. I also transferred the DHT support code to the second core of the Pico (core 1), so that the callback's and interrupts don't interfere with DHT communication. (but I think you didn't implement a DHT for outside temperature).

**Comment by UltrashRicco** [comment]
Hi André, thank you for your time!
I did not implement a DHT for outside temperature indeed.
I could remove the RTC battery prior to performing any test. Please let me know whether you have  a test protocol to suggest, and I would gladly follow it.

**Reply by astlouys** [reply]
Hi Eric,

I just released Version 9.00 of the Green Clock Firmware.

NTP has been implemented (of course, it requires a Pico W) and I hope it won't have side effects with the Summer Time handling. It is now more difficult to make tests with Summer Timer since NTP brings back local time relatively fast. I will keep an eye next spring, when time change occurs.

I will mark this discussion as "closed" since Version 6.00 (in this thread title) is far behind ;-)

André
